{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Environment: This script should be run with the `python_scvi_environment` environment using the devcontainer `docker_python_scvi`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "from scipy.stats import spearmanr\n",
    "from scvi.data import cortex, smfish\n",
    "from scvi.external import GIMVI\n",
    "import os\n",
    "import pandas as pd\n",
    "from sklearn.neighbors import KDTree\n",
    "from tqdm.notebook import tqdm\n",
    "from collections import Counter\n",
    "from sklearn.neighbors import KDTree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_size = 1\n",
    "\n",
    "%config InlineBackend.print_figure_kwargs={'facecolor' : \"w\"}\n",
    "%config InlineBackend.figure_format='retina'\n",
    "# set scvi seed\n",
    "np.random.seed(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = \"../../data\"\n",
    "output_folder = \"../../outputs\"\n",
    "\n",
    "try:\n",
    "    os.mkdir(os.path.join(data_path, \"integration\"))\n",
    "except FileExistsError:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = os.path.join(output_folder, \"adatas\", \"joint_embedded_R_for_transfer.h5ad\")\n",
    "spatial_data = sc.read(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_file = os.path.join(data_path, \"temp_objects\", \"AvrRpt2_alone2.h5ad\")\n",
    "seq_data = sc.read(seq_file)\n",
    "seq_data = seq_data[\n",
    "    seq_data.obs[\"sample.order\"].isin(\n",
    "        [\n",
    "            \"00_00_Mock_rep1\",\n",
    "            \"02_AvrRpt2_04h_rep1\",\n",
    "            \"02_AvrRpt2_06h_rep1\",\n",
    "            \"02_AvrRpt2_09h_rep1\",\n",
    "            \"02_AvrRpt2_24h_rep1\",\n",
    "        ]\n",
    "    )\n",
    "]\n",
    "seq_data.obs[\"sample.order\"] = (\n",
    "    seq_data.obs[\"sample.order\"]\n",
    "    .replace(\"00_00_Mock_rep1\", \"mock\")\n",
    "    .replace(\"02_AvrRpt2_04h_rep1\", \"4hr_avr\")\n",
    "    .replace(\"02_AvrRpt2_06h_rep1\", \"6hr_avr\")\n",
    "    .replace(\"02_AvrRpt2_09h_rep1\", \"9hr_avr\")\n",
    "    .replace(\"02_AvrRpt2_24h_rep1\", \"avrrpt24\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersection = []\n",
    "non_intersecting = []\n",
    "for k in spatial_data.var_names:\n",
    "    if k in seq_data.var_names:\n",
    "        intersection.append(k)\n",
    "    else:\n",
    "        non_intersecting.append(k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# only use genes in both datasets\n",
    "seq_data = seq_data[:, intersection].copy()\n",
    "spatial_data = spatial_data[:, intersection].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def index_nomenclature_alignment(index):\n",
    "    return [\n",
    "        i.replace(\"00_col_Mock_rep1\", \"00_Mock\")\n",
    "        .replace(\"col_AvrRpt2_24h_rep1\", \"AvrRpt2_24h\")\n",
    "        .replace(\"col_AvrRpt2_04h_rep1\", \"AvrRpt2_04h\")\n",
    "        .replace(\"col_AvrRpt2_06h_rep1\", \"AvrRpt2_06h\")\n",
    "        .replace(\"col_AvrRpt2_09h_rep1\", \"AvrRpt2_09h\")\n",
    "        for i in index\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_pseudotime_to_seq_data(pseudotime_path, seq_data):\n",
    "    rep1_pseudotime = pd.read_csv(pseudotime_path)\n",
    "\n",
    "    seq_data_with_pseudotime = seq_data.copy()\n",
    "    seq_data_with_pseudotime.obs.index = index_nomenclature_alignment(\n",
    "        seq_data_with_pseudotime.obs.index\n",
    "    )\n",
    "\n",
    "    new_df = seq_data_with_pseudotime.obs.merge(\n",
    "        rep1_pseudotime, how=\"left\", left_index=True, right_on=\"index\"\n",
    "    )\n",
    "    new_df.index = new_df[\"index\"]\n",
    "    new_df[\"dpt_pseudotime\"] = new_df[\"dpt_pseudotime\"].fillna(0)\n",
    "    seq_data_with_pseudotime.obs = new_df\n",
    "\n",
    "    return seq_data_with_pseudotime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pseudotime_path = os.path.join(data_path, \"temp_objects\", \"pseudotime_csv.csv\")\n",
    "\n",
    "seq_data_with_pseudotime = add_pseudotime_to_seq_data(pseudotime_path, seq_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_data_samples = [\n",
    "    \"00_Mock\",\n",
    "    \"AvrRpt2_04h\",\n",
    "    \"AvrRpt2_06h\",\n",
    "    \"AvrRpt2_09h\",\n",
    "    \"AvrRpt2_24h\",\n",
    "]\n",
    "spatial_data_samples = [\"mock\", \"4hr_avr\", \"6hr_avr\", \"9hr_avr\", \"avrrpt24\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_and_run_GIMVI(\n",
    "    current_seq_experiment,\n",
    "    current_spatial_experiment,\n",
    "    train_size=1,\n",
    "    n_epochs=200,\n",
    "    n_latent=10,\n",
    "):\n",
    "\n",
    "    seq_gene_names = current_seq_experiment.var_names\n",
    "    n_genes = current_seq_experiment.n_vars\n",
    "    n_train_genes = int(n_genes * train_size)\n",
    "\n",
    "    # randomly select training_genes\n",
    "    rand_train_gene_idx = np.random.choice(range(n_genes), n_train_genes, replace=False)\n",
    "    rand_test_gene_idx = sorted(set(range(n_genes)) - set(rand_train_gene_idx))\n",
    "    rand_train_genes = seq_gene_names[rand_train_gene_idx]\n",
    "    rand_test_genes = seq_gene_names[rand_test_gene_idx]\n",
    "\n",
    "    # current_spatial_experiment_partial has a subset of the genes to train on\n",
    "    # current_spatial_experiment_partial = current_spatial_experiment[:, rand_train_genes].copy()\n",
    "\n",
    "    # remove cells with no counts\n",
    "    sc.pp.filter_cells(current_spatial_experiment, min_counts=1)\n",
    "    sc.pp.filter_cells(current_seq_experiment, min_counts=1)\n",
    "\n",
    "    # setup_anndata for spatial and sequencing data\n",
    "    GIMVI.setup_anndata(\n",
    "        current_spatial_experiment, labels_key=\"leiden\", batch_key=\"batch\"\n",
    "    )\n",
    "    GIMVI.setup_anndata(\n",
    "        current_seq_experiment, labels_key=\"SCT_snn_res.1\", batch_key=\"sample.order\"\n",
    "    )\n",
    "\n",
    "    current_spatial_experiment.X = current_spatial_experiment.layers[\"counts\"]\n",
    "    # create our model\n",
    "    model = GIMVI(current_seq_experiment, current_spatial_experiment, n_latent=10)\n",
    "\n",
    "    # train for 200 epochs\n",
    "    model.train(200)\n",
    "\n",
    "    # get the latent representations for the sequencing and spatial data\n",
    "    latent_seq, latent_spatial = model.get_latent_representation()\n",
    "\n",
    "    # concatenate to one latent representation\n",
    "    latent_representation = np.concatenate([latent_seq, latent_spatial])\n",
    "    latent_adata = anndata.AnnData(latent_representation)\n",
    "\n",
    "    # labels which cells were from the sequencing dataset and which were from the spatial dataset\n",
    "    latent_labels = ([\"seq\"] * latent_seq.shape[0]) + (\n",
    "        [\"spatial\"] * latent_spatial.shape[0]\n",
    "    )\n",
    "    latent_adata.obs[\"labels\"] = latent_labels\n",
    "\n",
    "    # compute umap\n",
    "    sc.pp.neighbors(latent_adata, use_rep=\"X\")\n",
    "    sc.tl.umap(latent_adata)\n",
    "\n",
    "    # save umap representations to original seq and current_spatial_experimentsets\n",
    "    current_seq_experiment.obsm[\"X_umap_gimvi\"] = latent_adata.obsm[\"X_umap\"][\n",
    "        : current_seq_experiment.shape[0]\n",
    "    ]\n",
    "    current_spatial_experiment.obsm[\"X_umap_gimvi\"] = latent_adata.obsm[\"X_umap\"][\n",
    "        current_seq_experiment.shape[0] :\n",
    "    ]\n",
    "\n",
    "    seq_latent = model.get_latent_representation()[0]\n",
    "\n",
    "    current_seq_experiment.obsm[\"X_gimvi\"] = seq_latent\n",
    "    current_spatial_experiment.obsm[\"X_gimvi\"] = model.get_latent_representation()[1]\n",
    "\n",
    "    sc.pp.neighbors(latent_adata, use_rep=\"X\", n_neighbors=30)\n",
    "    sc.tl.umap(latent_adata, min_dist=0.1)\n",
    "\n",
    "    # save umap representations to original seq and current_spatial_experimentsets\n",
    "    current_seq_experiment.obsm[\"X_umap_gimvi\"] = latent_adata.obsm[\"X_umap\"][\n",
    "        : current_seq_experiment.shape[0]\n",
    "    ]\n",
    "    current_spatial_experiment.obsm[\"X_umap_gimvi\"] = latent_adata.obsm[\"X_umap\"][\n",
    "        current_seq_experiment.shape[0] :\n",
    "    ]\n",
    "\n",
    "    sc.pl.embedding(latent_adata, basis=\"X_umap\", color=\"labels\")\n",
    "\n",
    "    return current_seq_experiment, current_spatial_experiment, latent_adata, model\n",
    "\n",
    "\n",
    "def impute_pseudotime(\n",
    "    latent_adata, single_experiment_seq_data, current_spatial_experiment, n_neighbors=10\n",
    "):\n",
    "\n",
    "    # Assuming your original AnnData object is named 'adata'\n",
    "    adata = latent_adata.copy()\n",
    "\n",
    "    # Step 1: Filter spatial and seq cells\n",
    "    spatial_cells = adata[adata.obs[\"labels\"] == \"spatial\"].copy()\n",
    "    seq_cells = adata[adata.obs[\"labels\"] == \"seq\"].copy()\n",
    "    single_experiment_seq_data_copy = single_experiment_seq_data.copy()\n",
    "\n",
    "    # Step 2: Create KD tree for seq cells\n",
    "    seq_gimvi = single_experiment_seq_data.obsm[\"X_gimvi\"]\n",
    "    kdtree = KDTree(seq_gimvi)\n",
    "\n",
    "    # Step 3: Find nearest seq cells for each seq\n",
    "    distances, indices = kdtree.query(\n",
    "        current_spatial_experiment.obsm[\"X_gimvi\"], k=n_neighbors\n",
    "    )\n",
    "\n",
    "    # Step 4: Average gene expression for spatial cells\n",
    "    averaged_expression = np.zeros((spatial_cells.n_obs, 1))\n",
    "\n",
    "    for i in tqdm(range(spatial_cells.n_obs)):\n",
    "        seq_neighbors_indices = indices[i]\n",
    "        seq_neighbors_expression = single_experiment_seq_data_copy.obs[\n",
    "            \"dpt_pseudotime\"\n",
    "        ].values[seq_neighbors_indices]\n",
    "\n",
    "        averaged_expression[i] = np.mean(seq_neighbors_expression, axis=0)\n",
    "\n",
    "    pseudotime = averaged_expression.flatten()\n",
    "\n",
    "    return pseudotime\n",
    "\n",
    "\n",
    "def impute_celltype(\n",
    "    latent_adata, single_experiment_seq_data, spatial_data, n_neighbors=10\n",
    "):\n",
    "    # Assuming your original AnnData object is named 'adata'\n",
    "    adata = latent_adata.copy()\n",
    "\n",
    "    # Step 1: Filter spatial and seq cells\n",
    "    spatial_cells = adata[adata.obs[\"labels\"] == \"spatial\"].copy()\n",
    "    seq_cells = adata[adata.obs[\"labels\"] == \"seq\"].copy()\n",
    "    single_experiment_seq_data_copy = single_experiment_seq_data.copy()\n",
    "\n",
    "    # Step 2: Create KD tree for seq cells\n",
    "    seq_gimvi = single_experiment_seq_data.obsm[\"X_gimvi\"]\n",
    "    kdtree = KDTree(seq_gimvi)\n",
    "\n",
    "    # Step 3: Find nearest seq cells for each seq\n",
    "    n_neighbors = 30  # Specify the number of nearest neighbors\n",
    "    distances, indices = kdtree.query(spatial_data.obsm[\"X_gimvi\"], k=n_neighbors)\n",
    "\n",
    "    # Step 4: Average gene expression for spatial cells\n",
    "    averaged_expression = np.zeros((spatial_cells.n_obs, 1))\n",
    "    average_celltype = []\n",
    "\n",
    "    for i in tqdm(range(spatial_cells.n_obs)):\n",
    "        seq_neighbors_indices = indices[i]\n",
    "        seq_neighbors_categories = single_experiment_seq_data_copy.obs[\n",
    "            \"SCT_snn_res.1\"\n",
    "        ].values[seq_neighbors_indices]\n",
    "\n",
    "        # Use Counter to count occurrences\n",
    "        counter = Counter(seq_neighbors_categories)\n",
    "\n",
    "        # Find the most common element\n",
    "        most_common = counter.most_common(1)\n",
    "\n",
    "        averaged_expression[i] = most_common[0][0]\n",
    "\n",
    "        seq_neighbors_categories = single_experiment_seq_data_copy.obs[\n",
    "            \"celltype\"\n",
    "        ].values[seq_neighbors_indices]\n",
    "\n",
    "        # Use Counter to count occurrences\n",
    "        counter = Counter(seq_neighbors_categories)\n",
    "\n",
    "        # Find the most common element\n",
    "        most_common = counter.most_common(1)\n",
    "\n",
    "        average_celltype.append(most_common[0][0])\n",
    "\n",
    "    clustering = averaged_expression.flatten().astype(int).astype(str)\n",
    "\n",
    "    return clustering, average_celltype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for samp in range(len(seq_data_samples)):\n",
    "    print(samp)\n",
    "    current_seq_sample = seq_data_samples[samp]\n",
    "    current_spatial_sample = spatial_data_samples[samp]\n",
    "\n",
    "    current_seq_experiment = seq_data[seq_data.obs[\"sample2\"] == current_seq_sample]\n",
    "    current_spatial_experiment = spatial_data[\n",
    "        spatial_data.obs[\"batch\"] == current_spatial_sample\n",
    "    ]\n",
    "\n",
    "    current_seq_experiment, current_spatial_experiment, latent_adata, model = (\n",
    "        prepare_and_run_GIMVI(\n",
    "            current_seq_experiment,\n",
    "            current_spatial_experiment,\n",
    "            train_size=train_size,\n",
    "            n_epochs=200,\n",
    "            n_latent=10,\n",
    "        )\n",
    "    )\n",
    "\n",
    "    single_experiment_seq_data = seq_data_with_pseudotime[\n",
    "        seq_data_with_pseudotime.obs[\"sample2\"] == current_seq_sample\n",
    "    ]\n",
    "\n",
    "    current_seq_experiment.obs.index = index_nomenclature_alignment(\n",
    "        current_seq_experiment.obs.index\n",
    "    )\n",
    "\n",
    "    current_seq_experiment = current_seq_experiment[\n",
    "        current_seq_experiment.obs.index.isin(single_experiment_seq_data.obs.index)\n",
    "    ]\n",
    "\n",
    "    single_experiment_seq_data = single_experiment_seq_data[\n",
    "        single_experiment_seq_data.obs.index.isin(current_seq_experiment.obs.index)\n",
    "    ]\n",
    "\n",
    "    single_experiment_seq_data.obsm[\"X_gimvi\"] = current_seq_experiment.obsm[\"X_gimvi\"]\n",
    "\n",
    "    single_experiment_seq_data.obsm[\"X_umap_gimvi\"] = current_seq_experiment.obsm[\n",
    "        \"X_umap_gimvi\"\n",
    "    ]\n",
    "\n",
    "    current_spatial_experiment.obs[\"pseudotime\"] = impute_pseudotime(\n",
    "        latent_adata,\n",
    "        single_experiment_seq_data,\n",
    "        current_spatial_experiment,\n",
    "        n_neighbors=10,\n",
    "    )\n",
    "\n",
    "    clustering, average_celltype = impute_celltype(\n",
    "        latent_adata,\n",
    "        single_experiment_seq_data,\n",
    "        current_spatial_experiment,\n",
    "        n_neighbors=20,\n",
    "    )\n",
    "    current_spatial_experiment.obs[\"SCT_snn_res.1\"] = clustering\n",
    "    current_spatial_experiment.obs[\"celltype\"] = average_celltype\n",
    "\n",
    "    curr_directory = os.path.join(data_path, \"integration\", current_spatial_sample)\n",
    "\n",
    "    try:\n",
    "        os.mkdir(curr_directory)\n",
    "    except FileExistsError:\n",
    "        pass\n",
    "\n",
    "    latent_adata.write(os.path.join(curr_directory, \"latent_adata.h5ad\"))\n",
    "    current_spatial_experiment.write(os.path.join(curr_directory, \"spatial_data.h5ad\"))\n",
    "    single_experiment_seq_data.write(os.path.join(curr_directory, \"seq_data.h5ad\"))\n",
    "\n",
    "    sc.pl.embedding(\n",
    "        current_spatial_experiment, basis=\"umap_gimvi\", color=[\"celltype\", \"pseudotime\"]\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scvi-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
